Java/core/src/main/java/com/amazon/randomcutforest/store/SmallIndexManager.java

```diff
@@ -125,8 +182,26 @@ protected short takeIndex() {
      * @param index The index value to release.
      */
     protected void releaseIndex(int index) {
-        checkValidIndex(index);
+        checkArgument(index >= 0 && index < capacity, " incorrect index to release");
+        checkArgument(occupied.get(index), " releasing the index twice");
         occupied.clear(index);
+        if (freeIndexPointer + 1 >= freeIndexes.length) {
+            if (index == capacity - (freeIndexPointer + 1) - 1) {
+                ++freeIndexPointer;
+                return;
+            } else {
+                // for most of the applications of this small sampler we would need one extra
+                // entry; but freeIndexPointer can be -1 as well
+                // this can be changed to a larger value to reduce repeated
+                // allocation/deallocation
+                int cap = Math.min(capacity, freeIndexPointer + 2);
+                int oldLength = freeIndexes.length;
+                freeIndexes = Arrays.copyOf(freeIndexes, cap);
```

Comment:
Now that we're doing dynamic array resizing, I wonder if it would make sense to just use a Java List or something as a stack, and then we don't have to write the bookkeeping code.

Reply:
possible. Lets switch once the other changes are in.